File: sorting/Radix_sort.c
#include <stdio.h>
#include <stdlib.h>

int get_max(int A[], int n)
{
    int max = A[0];
    for (int i = 1; i < n; i++)
    {
        if (A[i] > max)
            max = A[i];
    }
    return max;
}

void count_sort(int A[], int n, int pos)
{
    int output[n];       
    int count[10] = {0};

    for (int i = 0; i < n; i++)
        count[(A[i] / pos) % 10]++;

    for (int i = 1; i < 10; i++)
        count[i] += count[i - 1];

    for (int i = n - 1; i >= 0; i--)
    {
        int digit = (A[i] / pos) % 10;
        output[--count[digit]] = A[i];
    }

    for (int i = 0; i < n; i++)
        A[i] = output[i];
}

void radix_sort(int A[], int n)
{
    int max = get_max(A, n);
    for (int pos = 1; max / pos > 0; pos *= 10)
        count_sort(A, n, pos);
}

int main()
{
    int A[] = {170, 45, 75, 90, 802, 24, 2, 66};
    int n = sizeof(A) / sizeof(A[0]);

    radix_sort(A, n);

    printf("Sorted array: ");
    for (int i = 0; i < n; i++)
        printf("%d ", A[i]);
    printf("\n");

    return 0;
}

File: sorting/radix_sort.py
def count_sort(arr, pos):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for i in range(0,n):
        temp= (arr[i] // pos) % 10
        count[temp] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]
        
    i=n-1
    while i>=0:
        temp=(arr[i]//pos)%10
        output[count[temp]-1]=arr[i]
        count[temp] -= 1
        i =i-1

    for i in range(0,n):
        arr[i] = output[i]


def radix_sort(arr):
    max_val = max(arr)
    #for unit place pos=1
    pos = 1
    while max_val // pos > 0:
        count_sort(arr, pos)
        pos *= 10

arr = [157,15,7,20,1,72,333,111,605,225]
print("The original array is:", arr)
radix_sort(arr)
print("The Sorted array is:", arr)
